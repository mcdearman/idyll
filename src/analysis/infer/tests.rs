use crate::{
    analysis::infer::TypeSolver,
    lex::{token::Token, token_iter::TokenIter},
    parse::parse,
    rename::resolver::Resolver,
};
use logos::Logos;

fn test_helper(src: &str) {
    let lexer = TokenIter::new(src);
    let (ast, errors) = parse(lexer, true);
    if !errors.is_empty() {
        for e in errors {
            eprintln!("{}", e);
        }
        panic!("parsing failed");
    }
    let mut res = Resolver::new();
    let (nir, errors) = res.resolve(&ast.unwrap());
    if !errors.is_empty() {
        for e in errors {
            eprintln!("{}", e);
        }
        panic!("resolution failed");
    }
    let nir = nir.unwrap();
    let builtins = res.builtins();
    let scoped_interner = res.env().dump_to_interner();
    if !errors.is_empty() {
        for e in errors {
            eprintln!("{}", e);
        }
        panic!("resolution failed");
    }
    let mut solver = TypeSolver::new(builtins.clone(), scoped_interner);
    let (tir, errors) = solver.infer(src, &nir);
    if !errors.is_empty() {
        for e in errors {
            eprintln!("{}", e);
        }
        panic!("type inference failed");
    }
    let tir = tir.unwrap();
    insta::assert_debug_snapshot!(tir);
}

#[test]
fn test_infer_int() {
    test_helper("42");
}

#[test]
fn test_infer_let_decl() {
    test_helper("let x = 42");
}
