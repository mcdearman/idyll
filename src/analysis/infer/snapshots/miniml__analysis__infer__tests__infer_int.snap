---
source: src/analysis/infer/tests.rs
expression: "test_helper(\"42\")"
---
Root {
    decls: [
        Decl {
            kind: Fn(
                ScopedIdent {
                    id: @16,
                    key: InternedString(main),
                    span: 0..2,
                },
                [
                    Pattern {
                        kind: Ident(
                            ScopedIdent {
                                id: @17,
                                key: InternedString(args),
                                span: 0..2,
                            },
                        ),
                        ty: @49,
                        span: 0..2,
                    },
                ],
                Expr {
                    kind: Lit(
                        Int(42),
                    ),
                    ty: Int,
                    span: 0..2,
                },
            ),
            ty: @49 -> Int,
            span: 0..2,
        },
    ],
    span: 0..2,
}
